### ISSUE TITLE
missing dependency "Could not find meson"

### ISSUE THREAD

**Issue by beenotung** [deltachat/deltachat-desktop#784]
- Platform (linux/mac/windows): archlinux (awesome wm)
- Delta Chat Desktop Version: commit: 7528400eaff378f5fa1cdd8067a1a56415a1ad6e
- Expected behavior: should build without error
- Actual behavior: see error message of "Could not find meson"
- Steps to reproduce the problem:
```bash
git clone https://github.com/deltachat/deltachat-desktop
cd deltachat-desktop
# optional 
# git checkout 7528400eaff378f5fa1cdd8067a1a56415a1ad6e
npm i
npm run build
```
- Screenshots:
![1](https://user-images.githubusercontent.com/6007292/56956480-df68a180-6b76-11e9-91ce-683f8d4e28e3.png)

- Logs:
```
npm i

> electron-chromedriver@3.0.0 install /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/electron-chromedriver
> node ./download-chromedriver.js


> deltachat-node@0.43.0 install /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> npm run rebuild-all


> deltachat-node@0.43.0 rebuild-all /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> npm run rebuild-core && npm run rebuild-bindings


> deltachat-node@0.43.0 rebuild-core /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> node scripts/rebuild-core.js

Could not find meson
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-core: `node scripts/rebuild-core.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-core script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_22_986Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-all: `npm run rebuild-core && npm run rebuild-bindings`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-all script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_22_996Z-debug.log
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.8 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.8: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 install: `npm run rebuild-all`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_24_143Z-debug.log
npm i  19.97s user 3.05s system 130% cpu 17.621 total
[18:32] beenotung:deltachat-desktop$ npm i -D meson

> deltachat-node@0.43.0 install /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> npm run rebuild-all


> deltachat-node@0.43.0 rebuild-all /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> npm run rebuild-core && npm run rebuild-bindings


> deltachat-node@0.43.0 rebuild-core /home/beenotung/workspace/github.com/deltachat/deltachat-desktop/node_modules/deltachat-node
> node scripts/rebuild-core.js

Could not find meson
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-core: `node scripts/rebuild-core.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-core script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_46_025Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-all: `npm run rebuild-core && npm run rebuild-bindings`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-all script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_46_036Z-debug.log
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.8 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.8: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 install: `npm run rebuild-all`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T10_32_46_690Z-debug.log

```

**Comment by ralphtheninja** [comment]
At the moment `meson` is a required dependency when building from source, see Troubleshooting section https://github.com/deltachat/deltachat-desktop#troubleshooting

**Comment by beenotung** [comment]
then it should be installed by the archlinux package

**Comment by beenotung** [comment]
still got error after installing meson:
![1](https://user-images.githubusercontent.com/6007292/56980519-27a3b600-6baf-11e9-893b-efa8c6357263.png)

console log:
```
:: Parsing SRCINFO (1/1): deltachat-desktop-git
==> Making package: deltachat-desktop-git v0.98.2.r10.g8a456f7-1 (Wednesday, May 01, 2019 AM01:15:53 HKT)
==> Retrieving sources...
  -> Cloning deltachat-desktop-git git repo...
Cloning into bare repository '/home/beenotung/.cache/yay/deltachat-desktop-git/deltachat-desktop-git'...
remote: Enumerating objects: 165, done.
remote: Counting objects: 100% (165/165), done.
remote: Compressing objects: 100% (104/104), done.
remote: Total 10418 (delta 105), reused 95 (delta 60), pack-reused 10253
Receiving objects: 100% (10418/10418), 10.91 MiB | 694.00 KiB/s, done.
Resolving deltas: 100% (7565/7565), done.
  -> Found deltachat-desktop.desktop
  -> Found deltachat-desktop.sh
==> Validating source files with sha256sums...
    deltachat-desktop-git ... Skipped
    deltachat-desktop.desktop ... Passed
    deltachat-desktop.sh ... Passed
==> Cleaning up...
==> Making package: deltachat-desktop-git v0.98.2.r10.g8a456f7-1 (Wednesday, May 01, 2019 AM01:16:12 HKT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Updating deltachat-desktop-git git repo...
Fetching origin
  -> Found deltachat-desktop.desktop
  -> Found deltachat-desktop.sh
==> Validating source files with sha256sums...
    deltachat-desktop-git ... Skipped
    deltachat-desktop.desktop ... Passed
    deltachat-desktop.sh ... Passed
==> Removing existing $srcdir/ directory...
==> Extracting sources...
  -> Creating working copy of deltachat-desktop-git git repo...
Cloning into 'deltachat-desktop-git'...
done.
==> Starting prepare()...

> electron-chromedriver@3.0.0 install /home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/electron-chromedriver
> node ./download-chromedriver.js


> deltachat-node@0.43.0 install /home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node
> npm run rebuild-all


> deltachat-node@0.43.0 rebuild-all /home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node
> npm run rebuild-core && npm run rebuild-bindings


> deltachat-node@0.43.0 rebuild-core /home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node
> node scripts/rebuild-core.js

[140/168] Compiling C object 'src/25a6634@@deltachat@sta/dc_imex.c.o'.
../src/dc_imex.c: In function ‘dc_create_setup_code’:
../src/dc_imex.c:261:5: warning: ‘RAND_pseudo_bytes’ is deprecated [-Wdeprecated-declarations]
     RAND_pseudo_bytes((unsigned char*)&random_val, sizeof(uint16_t));
     ^~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/ossl_typ.h:19,
                 from /usr/include/openssl/rand.h:14,
                 from ../src/dc_imex.c:4:
/usr/include/openssl/rand.h:44:1: note: declared here
 DEPRECATEDIN_1_1_0(int RAND_pseudo_bytes(unsigned char *buf, int num))
 ^~~~~~~~~~~~~~~~~~
[150/168] Compiling C object 'src/25a6634@@deltachat@sta/dc_openssl.c.o'.
../src/dc_openssl.c:85:13: warning: ‘dyn_destroy_function’ defined but not used [-Wunused-function]
 static void dyn_destroy_function(struct CRYPTO_dynlock_value* l, const char* file, int line)
             ^~~~~~~~~~~~~~~~~~~~
../src/dc_openssl.c:74:13: warning: ‘dyn_lock_function’ defined but not used [-Wunused-function]
 static void dyn_lock_function(int mode, struct CRYPTO_dynlock_value* l, const char* file, int line)
             ^~~~~~~~~~~~~~~~~
../src/dc_openssl.c:63:37: warning: ‘dyn_create_function’ defined but not used [-Wunused-function]
 static struct CRYPTO_dynlock_value* dyn_create_function(const char* file, int line)
                                     ^~~~~~~~~~~~~~~~~~~
../src/dc_openssl.c:46:13: warning: ‘locking_function’ defined but not used [-Wunused-function]
 static void locking_function(int mode, int n, const char* file, int line)
             ^~~~~~~~~~~~~~~~
../src/dc_openssl.c:40:22: warning: ‘id_function’ defined but not used [-Wunused-function]
 static unsigned long id_function(void)
                      ^~~~~~~~~~~
[164/168] Compiling C object 'src/25a6634@@deltachat@sta/dc_tools.c.o'.
../src/dc_tools.c: In function ‘dc_create_id’:
../src/dc_tools.c:963:3: warning: ‘RAND_pseudo_bytes’ is deprecated [-Wdeprecated-declarations]
   RAND_pseudo_bytes((unsigned char*)&buf, sizeof(uint32_t)*3);
   ^~~~~~~~~~~~~~~~~
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/ossl_typ.h:19,
                 from /usr/include/openssl/rand.h:14,
                 from ../src/dc_tools.c:8:
/usr/include/openssl/rand.h:44:1: note: declared here
 DEPRECATEDIN_1_1_0(int RAND_pseudo_bytes(unsigned char *buf, int num))
 ^~~~~~~~~~~~~~~~~~
[168/168] Linking target cmdline/delta.
make: Entering directory '/home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node/build'
  CC(target) Release/obj.target/deltachat/src/module.o
../src/module.c:23:3: error: unknown type name ‘napi_threadsafe_function’
   napi_threadsafe_function threadsafe_event_handler;
   ^~~~~~~~~~~~~~~~~~~~~~~~
../src/module.c: In function ‘dc_event_handler’:
../src/module.c:84:9: warning: implicit declaration of function ‘napi_call_threadsafe_function’; did you mean ‘napi_create_function’? [-Wimplicit-function-declaration]
         napi_call_threadsafe_function(dcn_context->threadsafe_event_handler, dcn_event, napi_tsfn_blocking);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         napi_create_function
../src/module.c:84:89: error: ‘napi_tsfn_blocking’ undeclared (first use in this function); did you mean ‘napi_run_script’?
         napi_call_threadsafe_function(dcn_context->threadsafe_event_handler, dcn_event, napi_tsfn_blocking);
                                                                                         ^~~~~~~~~~~~~~~~~~
                                                                                         napi_run_script
../src/module.c:84:89: note: each undeclared identifier is reported only once for each function it appears in
../src/module.c: In function ‘imap_thread_func’:
../src/module.c:159:3: warning: implicit declaration of function ‘napi_acquire_threadsafe_function’; did you mean ‘napi_create_function’? [-Wimplicit-function-declaration]
   napi_acquire_threadsafe_function(dcn_context->threadsafe_event_handler);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   napi_create_function
../src/module.c:167:3: warning: implicit declaration of function ‘napi_release_threadsafe_function’; did you mean ‘napi_create_function’? [-Wimplicit-function-declaration]
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   napi_create_function
../src/module.c:167:75: error: ‘napi_tsfn_release’ undeclared (first use in this function); did you mean ‘napi_is_promise’?
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
                                                                           ^~~~~~~~~~~~~~~~~
                                                                           napi_is_promise
../src/module.c: In function ‘smtp_thread_func’:
../src/module.c:182:75: error: ‘napi_tsfn_release’ undeclared (first use in this function); did you mean ‘napi_is_promise’?
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
                                                                           ^~~~~~~~~~~~~~~~~
                                                                           napi_is_promise
../src/module.c: In function ‘mvbox_thread_func’:
../src/module.c:198:75: error: ‘napi_tsfn_release’ undeclared (first use in this function); did you mean ‘napi_is_promise’?
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
                                                                           ^~~~~~~~~~~~~~~~~
                                                                           napi_is_promise
../src/module.c: In function ‘sentbox_thread_func’:
../src/module.c:213:75: error: ‘napi_tsfn_release’ undeclared (first use in this function); did you mean ‘napi_is_promise’?
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
                                                                           ^~~~~~~~~~~~~~~~~
                                                                           napi_is_promise
../src/module.c: In function ‘dcn_context_new’:
../src/module.c:319:41: warning: assignment to ‘int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
   dcn_context->threadsafe_event_handler = NULL;
                                         ^
In file included from ../src/napi-macros-extensions.h:1,
                 from ../src/module.c:10:
../src/module.c: In function ‘dcn_set_event_handler’:
../src/module.c:1288:22: warning: implicit declaration of function ‘napi_create_threadsafe_function’; did you mean ‘napi_create_function’? [-Wimplicit-function-declaration]
   NAPI_STATUS_THROWS(napi_create_threadsafe_function(
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../napi-macros/napi-macros.h:21:8: note: in definition of macro ‘NAPI_STATUS_THROWS’
   if ((call) != napi_ok) { \
        ^~~~
../src/module.c: In function ‘dcn_unset_event_handler’:
../src/module.c:1402:75: error: ‘napi_tsfn_release’ undeclared (first use in this function); did you mean ‘napi_is_promise’?
   napi_release_threadsafe_function(dcn_context->threadsafe_event_handler, napi_tsfn_release);
                                                                           ^~~~~~~~~~~~~~~~~
                                                                           napi_is_promise
make: *** [deltachat.target.mk:106: Release/obj.target/deltachat/src/module.o] Error 1
make: Leaving directory '/home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/beenotung/.nvm/versions/node/v8.12.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Linux 5.0.9-arch1-1-ARCH
gyp ERR! command "/home/beenotung/.nvm/versions/node/v8.12.0/bin/node" "/home/beenotung/.nvm/versions/node/v8.12.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/beenotung/.cache/yay/deltachat-desktop-git/src/deltachat-desktop-git/node_modules/deltachat-node
gyp ERR! node -v v8.12.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
npx failed with code 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-core: `node scripts/rebuild-core.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-core script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T17_16_59_093Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 rebuild-all: `npm run rebuild-core && npm run rebuild-bindings`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 rebuild-all script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T17_16_59_203Z-debug.log
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.8 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.8: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@0.43.0 install: `npm run rebuild-all`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the deltachat-node@0.43.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/beenotung/.npm/_logs/2019-04-30T17_17_01_037Z-debug.log
==> ERROR: A failure occurred in prepare().
    Aborting...
Error making: deltachat-desktop-git
```

**Comment by ralphtheninja** [comment]
> then it should be installed by the archlinux package (aur/deltachat-desktop-git)

Maybe yes, I haven't personally been involved with that package. cc @Jikstra 

> still got error after installing meson (maybe this shall be forked into another issue):

There are multiple issues here. First of all you need node at least on version 10, i.e. current LTS version. Node 8 and below isn't supported. This explains the issue with `napi_create_threadsafe_function`.

**Comment by ralphtheninja** [comment]
The error with `RAND_pseudo_bytes` I _think_ is related to your ssl version. Again, deferring to @Jikstra  on what needs to be done.

**Comment by Jikstra** [comment]
Just updated the dependencies in the PKGBUILD: https://aur.archlinux.org/cgit/aur.git/commit/PKGBUILD?h=deltachat-desktop-git&id=21ad1220d0392554239c6bb020bb88111b4343a0

@beenotung can you check if this fixes at least the meson error? In general it also says that openssl is needed, but i have something in my mind that deltachat-core currently doesn't compile on a too new openssl.

**Comment by ralphtheninja** [comment]
@Jikstra Please don't depend on odd number version of node. I think we should say node 10, since that's the version we use to build with.

**Comment by Jikstra** [comment]
> @Jikstra Please don't depend on odd number version of node. I think we should say node 10, since that's the version we use to build with.

Shouldn't be a big problem as in the arch repos there's only one version of node, and that's the most recent one. Just put in the one that i currently have. Also it says `greater than equal` so it should be at least that version. Depending on 10 won't work as that will generate conflicts with other packages that need a more recent version.

**Comment by ralphtheninja** [comment]
> Depending on 10 won't work as that will generate conflicts with other packages that need a more recent version.

Fair enough :smile:

**Comment by Jikstra** [comment]
@beenotung any news on your side?

**Comment by Simon-Laux** [comment]
> @beenotung any news on your side?

@beenotung is this solved?

**Comment by ralphtheninja** [comment]
Re-open if that's the case.

**Comment by beenotung** [comment]
does it depends on node 10 only at build-time or also in run-time?

**Comment by beenotung** [comment]
it install and launch properly when using node 10
